crates/ty_python_semantic/src/ast_node_ref.rs

```diff
@@ -52,34 +53,38 @@ impl<T> AstNodeRef<T> {
     /// Dereferencing the `node` can result in undefined behavior if `parsed` isn't the
     /// [`ParsedModuleRef`] to which `node` belongs. It's the caller's responsibility to ensure that
     /// the invariant `node belongs to parsed` is upheld.
-    pub(super) unsafe fn new(parsed: ParsedModuleRef, node: &T) -> Self {
+    pub(super) unsafe fn new(module_ref: &ParsedModuleRef, node: &T) -> Self {
         Self {
-            parsed,
-            node: std::ptr::NonNull::from(node),
+            module_ptr: module_ref.module().as_ptr(),
+            index: node.node_index().clone(),
+            _node: PhantomData,
         }
     }
 
     /// Returns a reference to the wrapped node.
     ///
-    /// Note that this method will panic if the provided module is from a different file or Salsa revision
-    /// than the module this node was created with.
-    pub fn node<'ast>(&self, parsed: &'ast ParsedModuleRef) -> &'ast T {
-        debug_assert!(Arc::ptr_eq(self.parsed.as_arc(), parsed.as_arc()));
+    /// Note that this method will panic if the provided module is from a different file or Salsa
+    /// revision than the module this node was created with.
+    pub fn node<'ast>(&self, module_ref: &'ast ParsedModuleRef) -> &'ast T {
+        debug_assert_eq!(module_ref.module().as_ptr(), self.module_ptr);
 
-        // SAFETY: Holding on to `parsed` ensures that the AST to which `node` belongs is still
-        // alive and not moved.
-        unsafe { self.node.as_ref() }
+        // Note that the module pointer is guaranteed to be stable within the Salsa
+        // revision, so the file contents cannot have changed by the above assertion.
+        module_ref
+            .get_by_index(&self.index)
+            .try_into()
+            .ok()
+            .expect("AST indices should never change within the same revision")
     }
 }
 
-impl<T> std::fmt::Debug for AstNodeRef<T>
+impl<T> Debug for AstNodeRef<T>
 where
-    T: std::fmt::Debug,
+    T: Debug,
 {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.debug_tuple("AstNodeRef")
-            .field(self.node(&self.parsed))
-            .finish()
+        // Unfortunately we have no access to the AST here.
+        f.debug_tuple("AstNodeRef").field(&"_").finish()
```

Comment:
Could we store the `NodeKind` and `TextRange` in debug builds and print them here. That would at least give us some idea of what node this is.